components/function-controller/internal/controllers/serverless/build_resources.go

```diff
@@ -172,6 +189,118 @@ func (r *FunctionReconciler) buildService(log logr.Logger, instance *serverlessv
 			},
 		},
 	}
+}
+
+func (r *FunctionReconciler) buildInternalImageAddress(instance *serverlessv1alpha1.Function) string {
+	imageTag := r.calculateImageTag(instance)
+	return fmt.Sprintf("%s/%s-%s:%s", r.config.Docker.Address, instance.Namespace, instance.Name, imageTag)
+}
+
+func (r *FunctionReconciler) buildExternalImageAddress(instance *serverlessv1alpha1.Function) string {
+	imageTag := r.calculateImageTag(instance)
+	return fmt.Sprintf("%s/%s-%s:%s", r.config.Docker.ExternalAddress, instance.Namespace, instance.Name, imageTag)
+}
+
+func (r *FunctionReconciler) sanitizeDependencies(dependencies string) string {
+	result := "{}"
+	if strings.Trim(dependencies, " ") != "" {
+		result = dependencies
+	}
+
+	return result
+}
+
+func (r *FunctionReconciler) functionLabels(instance *serverlessv1alpha1.Function) map[string]string {
+	labels := make(map[string]string, len(instance.GetLabels())+3)
+	for key, value := range instance.GetLabels() {
+		labels[key] = value
+	}
+
+	labels[serverlessv1alpha1.FunctionNameLabel] = instance.Name
+	labels[serverlessv1alpha1.FunctionManagedByLabel] = "function-controller"
+	labels[serverlessv1alpha1.FunctionUUIDLabel] = string(instance.GetUID())
+
+	return labels
+}
+
+func (r *FunctionReconciler) serviceLabels(instance *serverlessv1alpha1.Function) map[string]string {
+	serviceLabels := r.functionLabels(instance)
+	serviceLabels[servingKnativeVisibilityLabel] = servingKnativeVisibilityLabelValue
+	return serviceLabels
+}
+
+func (r *FunctionReconciler) servicePodAnnotations(instance *serverlessv1alpha1.Function) map[string]string {
+	annotations := map[string]string{
+		autoscalingKnativeMinScaleAnn: "1",
+		autoscalingKnativeMaxScaleAnn: "1",
+	}
+	if instance.Spec.MinReplicas != nil {
+		annotations[autoscalingKnativeMinScaleAnn] = fmt.Sprintf("%d", *instance.Spec.MinReplicas)
+	}
+	if instance.Spec.MaxReplicas != nil {
+		annotations[autoscalingKnativeMaxScaleAnn] = fmt.Sprintf("%d", *instance.Spec.MaxReplicas)
+	}
+	return annotations
+}
+
+func (r *FunctionReconciler) servicePodLabels(log logr.Logger, instance *serverlessv1alpha1.Function, oldService *servingv1.Service) map[string]string {
+	functionLabels := r.functionLabels(instance)
+	bindingLabels := r.retrieveBindingLabels(log, instance, oldService)
+	podLabels := instance.Spec.PodLabels
+
+	if podLabels == nil || len(podLabels) == 0 {
+		for key, value := range bindingLabels {
+			functionLabels[key] = value
+		}
+		return functionLabels
+	}
+
+	for key, value := range functionLabels {
+		podLabels[key] = value
+	}
+	for key, value := range bindingLabels {
+		podLabels[key] = value
+	}
+
+	return podLabels
+}
+
+func (r *FunctionReconciler) retrieveBindingLabels(log logr.Logger, instance *serverlessv1alpha1.Function, oldService *servingv1.Service) map[string]string {
+	bindingAnnotation := instance.GetAnnotations()[serviceBindingUsagesTracingAnnotation]
+	functionLabels := r.retrieveBindingLabelsFromAnnotation(log, bindingAnnotation)
```

Comment:
Rather `bindingLabels`

Reply:
I changed name to `bindingLabelsFromFunction`.